### ISSUE TITLE
Compile all matmul shapes is SD3 model for AIE

### ISSUE THREAD

**Issue by newling** [nod-ai/iree-amd-aie#285]
Tracker task for the shapes we should support with direct codegen

2 compiler errors need resolving. The errors can be reproduced with square matmuls. 

### Error 1: m=n=k=2304 (=2048+256) bf16 operands f32 output. 

The failure is in `irrt-to-ipu`

The error message is 

```
// iree-compile: iree-amd-aie/third_party/mlir-air/mlir/lib/Conversion/AIRRtToIpuPass.cpp:551: void 
(anonymous namespace)::tileIllegalWrapDim(airrt::DmaMemcpyNdOp): Assertion 
`!(const_wrap % (AIE2_WRAP_UPPER_BOUND / 2)) && "Currently do not support remainder tiles"' failed.
```

@erwei-xilinx is aware of this:
https://teams.microsoft.com/l/message/19:meeting_Zjc5ZmZhM2EtZDcxZS00NzYxLTliYmQtNGFlNzY1MDJhNjMy@thread.v2/1713812572221?context=%7B%22contextType%22%3A%22chat%22%7D

### Error 2: m=n=k=2432 (=2048+256+128) bf16 operands f32 output. 
Failure in `air-label-scf-for-to-ping-pong`

Error message is quite generic, needs further understanding:
```
error: block with no terminator, has %40 = "air.wait_all"(%arg8) : (!air.async.token) -> !air.async.token
note: see current operation: %40 = "air.wait_all"(%arg8) : (!air.async.token) -> !air.async.token
```

Example input IR:
```
!lhs = tensor<2432x2432xbf16>
!rhs = tensor<2432x2432xbf16>
!out = tensor<2432x2432xf32>
func.func @matmul_32x32_32xf32_(%lhs : !lhs, %rhs : !rhs) -> !out {
  %init_acc = tensor.empty() : !out
  %c0_acc_type = arith.constant 0.0 : f32
  %acc = linalg.fill ins(%c0_acc_type : f32) outs(%init_acc : !out) -> !out
  %result = linalg.matmul ins(%lhs, %rhs: !lhs, !rhs) outs(%acc: !out) -> !out
  return %result: !out
}
```

**Comment by newling** [comment]
### Investigation of Error 2:

m=n=k=2304 does not fail compilation as early as m=n=k=2432 so I compare the IR generated by passes between these 2, to see where the 2432 case goes hors piste. 

I see the first big difference in the pass `'AIRSpecializeDmaBroadcast` where 2304 (the "good" case) generates a bunch of 

```
 %11 = affine.if #set()[%arg12, %arg13] -> !air.async.token {
  ... 
 }
 ```
 
 blocks. It seems suspicious to me that there are no equivalent blocks for the 2432 case, as 2432 is evenly less well divided by 2 than is 2304. 
 
 
 The next pass, DmaToChannel, sees the 2432 IR explode to 707 lines, while the "bad" 2304 case sees the IR size at 315. This again is suspicious: why is the IR with a less divisible tensor size and more elements smaller?

**Comment by erwei-xilinx** [comment]
I created a PR to MLIR-AIR which fixes the error 1. I was able to get the error 1's GEMM shape to pass locally (although using an older xrt built in Jan. 2024). However, when I try to land this in IREE-AMD-AIE, it fails in CI with an XRT issue:

```
[XRT] ERROR: Failed to allocate host memory buffer (mmap(len=10616832, prot=3, flags=8193, offset=4294967296) failed (err=11): Resource temporarily unavailable), make sure host bank is enabled (see xbutil configure --host-mem)
```

Does anyone know what that means? Pls see this PR for details: https://github.com/nod-ai/iree-amd-aie/pull/289

**Comment by erwei-xilinx** [comment]
> The current status is summarised in https://github.com/nod-ai/iree-amd-aie/pull/286.

Thanks for doing the sweep. This is really helpful, and it helped us identify the issue which is specific to how `air-split-l2-memref` pass handles 4x2 herds, as we found out that all hanging cases were with 4x2 herds.

This PR in AIR should fix this issue. I was able to get those shapes to pass the test locally: https://github.com/Xilinx/mlir-air/pull/562
